cookbook/testing/widget-test-introduction.md

```diff
@@ -0,0 +1,266 @@
+create a Widget that displays a `title` and `message`. 
+
+```dart
+class MyWidget extends StatelessWidget {
+  final String title;
+  final String message;
+
+  const MyWidget({
+    Key key,
+    @required this.title,
+    @required this.message,
+  }) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'Flutter Demo',
+      home: Scaffold(
+        appBar: AppBar(
+          title: Text(title),
+        ),
+        body: Center(
+          child: Text(message),
+        ),
+      ),
+    );
+  }
+}
+``` 
+
+### 3. Create a `testWidgets` test
+
+Now that we have a Widget to test, we can begin writing our first test! To get 
+started, we'll use the 
+[`testWidgets`](https://docs.flutter.io/flutter/flutter_test/testWidgets.html)
+function provided by the `flutter_test` package to define a test. The 
+`testWidgets` function will allow us to define a Widget test and will create a
+`WidgetTester` for us to work with.
+
+Our test will verify that `MyWidget` displays a given title and message.
+
+<!-- skip -->
+```dart
+void main() {
+  // Define a test. The TestWidgets function will also provide a WidgetTester
+  // for us to work with. The WidgetTester will allow us to build and interact 
+  // with Widgets in the test environment.   
+  testWidgets('MyWidget has a title and message', (WidgetTester tester) async {
+    // Test code will go here!
+  });
+}
+```
+
+### 4. Build the Widget using the `WidgetTester`
+
+Next, we'll want to build `MyWidget` inside the test environment. To do so, we
+can use the 
+[`pumpWidget`](https://docs.flutter.io/flutter/flutter_test/WidgetTester/pumpWidget.html) 
+method provided by the `WidgetTester`. The `pumpWidget` method will build and 
+render the Widget we provide.
+
+In this case, we'll create a `MyWidget` instance that displays "T" as the title
+and "M" as the message.
+
+<!-- skip -->
+```dart
+void main() {
+  testWidgets('MyWidget has a title and message', (WidgetTester tester) async {
+    // Create the Widget tell the tester to build it
+    await tester.pumpWidget(MyWidget(title: 'T', message: 'M'));
+  });
+}
+```
+
+#### Note
+
+After the initial call to `pumpWidget`, the `WidgetTester` provides additional 
+ways to rebuild the same Widget. This is useful if you're working with a 
+`StatefulWidget` or animations. 
+
+For example, if we tap a button, and this button calls `setState`, Flutter will 
+not automatically rebuild your Widget in the test environment. We need to use
+one of the following methods to ask Flutter to build our Widget once again. 
+
+  - [tester.pump()](https://docs.flutter.io/flutter/flutter_test/TestWidgetsFlutterBinding/pump.html) 
+  : Triggers a rebuild of the Widget after a given duration.
+  - [tester.pumpAndSetttle()](https://docs.flutter.io/flutter/flutter_test/TestWidgetsFlutterBinding/pump.html)
+  : Triggers a rebuild of the Widget after animations have completed.
```

Comment:
Unless you have a good reason otherwise, consider using the text from the DartDoc for this one: 

> Repeatedly calls `pump` with the given duration until there are no longer any frames scheduled. This essentially waits for all animations to complete.